File: Local-TFG/Procesamiento de datos/prueba-2/version-1/javascript/main.js
function createDataStructure(size) {
    const arr = [];
    for (let i = 0; i < size; i++) {
        arr.push(Math.floor(Math.random() * 1001));
    }
    return arr;
}

function calculateSum(data) {
    let total = 0;
    for (const num of data) {
        total += num;
    }
    return total;
}

function calculateMean(data) {
    return calculateSum(data) / data.length;
}

function calculateStd(data) {
    const mean = calculateMean(data);
    let squaredDiffSum = 0;
    for (const num of data) {
        squaredDiffSum += Math.pow(num - mean, 2);
    }
    return Math.sqrt(squaredDiffSum / data.length);
}

function getMemoryUsageJS() {
    if (performance.memory) {
        return Math.max(performance.memory.usedJSHeapSize / (1024 * 1024), 0);
    }
    return -1;
}

function doStatisticalAnalysis(size) {
    const metrics = {};
    const startTotal = performance.now();
    let maxMemory = 0;

    let start_op = performance.now();
    let data = createDataStructure(size);
    metrics['create'] = {
        time: performance.now() - start_op,
        memory: getMemoryUsageJS(),
    };

    maxMemory = Math.max(maxMemory, getMemoryUsageJS());

    const operations = [
        ['sum', calculateSum],
        ['mean', calculateMean],
        ['std', calculateStd]
    ];

    for (const [op_name, op_func] of operations) {
        start_op = performance.now();

        const mem_before = getMemoryUsageJS();
        op_func(data);
        const mem_after = getMemoryUsageJS();

        if (mem_after > maxMemory) maxMemory = mem_after;

        metrics[op_name] = {
            time: performance.now() - start_op,
            memory: Math.abs(mem_after - mem_before),
        };
    }

    metrics['output'] = {
        total_time: performance.now() - startTotal,
        memory_peak: maxMemory
    };

    displayResults(metrics);
}


function displayResults(results) {
    const output = document.getElementById("javascript-output");

    for (const [op, data] of Object.entries(results)) {
        if (op !== 'output') {
            const timeDiv = document.createElement("div");
            timeDiv.textContent =
                `${op.toUpperCase()} - Time: ${data.time.toFixed(2)} ms | RAM: ${data.memory.toFixed(2)} MB`;
            output.appendChild(timeDiv);
        }
    }

    const timeTotalDiv = document.createElement("div");
    timeTotalDiv.textContent =
        `Total ET: ${results.output.total_time.toFixed(2)} ms`;
    output.appendChild(timeTotalDiv);

    const memoryDiv = document.createElement("div");
    memoryDiv.textContent =
        `RAM Peak: ${results.output.memory_peak.toFixed(2)} MB`;
    output.appendChild(memoryDiv);
}

function runJsBenchmark() {
    window.clearCell("javascript-output");
    doStatisticalAnalysis(10_000_000);
}

File: Local-TFG/Procesamiento de datos/prueba-2/version-1/python/main.py
import time
import tracemalloc
import gc
import js  # type: ignore
from pyscript import display
import random


def create_data_structure(size):
    return [random.randint(0, 1000) for _ in range(size)]


def calculate_sum(data):
    return sum(data)


def calculate_mean(data):
    return sum(data) / len(data)


def calculate_std(data):
    mean = calculate_mean(data)
    return (sum((x - mean) ** 2 for x in data) / len(data)) ** 0.5


def get_memory_usage():
    current, peak = tracemalloc.get_traced_memory()
    return peak / (1024 * 1024)


def do_statistical_analysis(size):
    metrics = {}
    max_memory = 0
    start_total = time.perf_counter()

    gc.collect()
    tracemalloc.start()
    start_op = time.perf_counter()
    data = create_data_structure(size)
    create_time = (time.perf_counter() - start_op) * 1000
    create_memory = get_memory_usage()
    tracemalloc.stop()

    metrics['create'] = {
        'time': create_time,
        'memory': create_memory
    }
    max_memory = max(max_memory, create_memory)

    operations = [
        ('sum', calculate_sum, (data,)),
        ('mean', calculate_mean, (data,)),
        ('std', calculate_std, (data,))
    ]

    for op_name, op_func, args in operations:
        gc.collect()
        tracemalloc.start()
        start_op = time.perf_counter()
        result = op_func(*args)
        op_time = (time.perf_counter() - start_op) * 1000
        op_memory = get_memory_usage()
        tracemalloc.stop()

        metrics[op_name] = {
            'time': op_time,
            'memory': op_memory
        }
        max_memory = max(max_memory, op_memory)

    metrics['output'] = {
        'total_time': (time.perf_counter() - start_total) * 1000,
        'memory_peak': max_memory
    }

    display_results(metrics)


def display_results(results):
    for op, data in results.items():
        if op != 'output':
            display(
                f"{op.upper()} - Time: {data['time']:.2f} ms | RAM: {data['memory']:.2f} MB",
                target="pyscript-output"
            )
    display(
        f"Total ET: {results['output']['total_time']:.2f} ms", target="pyscript-output"
    )
    display(
        f"RAM Peak: {results['output']['memory_peak']:.2f} MB", target="pyscript-output"
    )


def run_py_benchmark(event):
    js.clearCell('pyscript-output')
    do_statistical_analysis(10_000_000)
